distributed/scheduler.py

```diff
@@ -566,7 +584,7 @@ def __init__(
         )
 
         self._actors = set()
-        self._has_what = set()
+        self._has_what = {}
```

Comment:
This feels a bit fragile and couples things I would rather have independent.

This assumption is used in the algorithm below when defining `iter(ws._has_what)` as the tasks which are potential victims for the rebalancing. (Note, that this does _not_ use the indirection via the property `has_what`)
If we are shooting for a last inserted, I believe I would prefer to make this explicit and pay for it, e.g. `sorted(ws._has_what, key=lambda ws: ws.last_update)` (whatever last_update would be). This would also allow for an easier modification of this condition.
However, it works and it is cheap. If my proposal is too expensive I can probably live with this.

Reply:
the key point of this PR is to reduce the big-O complexity towards the total number of keys on the cluster from O(n*log(n)) to O(1). ``sorted`` would bring it back to O(n*log(n)) .

Reply:
> O(n*log(n)) to O(1). sorted would bring it back to O(n*log(n)) .

What is N in this equation? My suggestion would indeed put the (worst case) complexity to N logN where N is the number of keys on a given worker, not the total number of keys on the cluster. While formally all keys could live on a single worker, this is an edge case which is highly unlikely and should imho not drive this implementation

If we ever want to move away from the "last inserted" policy we will need to touch this again


My actual issue is not necessarily the complexity or extensibility but I have rather that we rely on a dict to achieve this _last inserted_ policy. The data structure which comes to my mind for something like this would be a stack. In python, that's most easily achieved using a `deque`.
Relying on how the dict is sorted is an implementation detail of CPython. Afaik, they guarantee that dicts stay sorted since py3.6(?) but afaik they do not guarantee _how_ they keys are sorted.
Not only do we rely on it to stay this way, I feel that this knowledge makes the code a little harder to maintain.

Reply:
N is the total number of keys on the cluster. The docstring gets thoroughly into the details of the big-O equation.

> The data structure which comes to my mind for something like this would be a stack. In python, that's most easily achieved using a deque.

keys must be removed from the middle of the data structure under normal operations. A deque does not offer this.

> Relying on how the dict is sorted is an implementation detail of CPython. 

It is not. dicts became insertion-sorted as an implementation detail in Python 3.6, which was then elevated to formal API spec in Python 3.7. All alternative implementations must implement the same to be called Python.

> but afaik they do not guarantee how they keys are sorted.

Yes there is. dicts are sorted by insertion order.

Please refer to https://docs.python.org/3.7/whatsnew/3.7.html

> the insertion-order preservation nature of dict objects has been declared to be an official part of the Python language spec.

And to https://docs.python.org/3/reference/datamodel.html

> Dictionaries preserve insertion order, meaning that keys will be produced in the same order they were added sequentially over the dictionary. Replacing an existing key does not change the order, however removing a key and re-inserting it will add it to the end instead of keeping its old place.

>  I feel that this knowledge makes the code a little harder to maintain.

I am extensively documenting that we are relying on this feature, and I will add thorough unit tests to this same PR to prevent accidental regressions.

Reply:
> try to decouple normal operations from rebalancing.  That wouldn't be the first time we're actually tracking data twice. 

The alternative to using a dict is to write a set-like class that is internally implemented with a dict. But I found that would be overkill and actually make everything harder to read.

Reply:
> > At this point we're not dealing with all tasks but only tasks in memory on a worker.
>
> Wrong, you're dealing with all tasks on all workers that need to donate keys

How is that wrong? The `has_what` is literally defined as the subset of tasks in memory. There is no general way to relate this number to the total number of tasks. There may be graphs with millions of tasks but only one final result key in memory. Only when scattering data, there is a 1to1 relation but for all others, this is non trivial.

---

I do not necessarily want to argue this academically. I do get your point about how the numbers can be related to the total number and that results in a big O of N log N but I argue that is only a valid argument given certain assumptions. these assumptions are sometimes true and sometimes not. If they are true, the question is whether this is relevant or not and how this impacts reality is difficult to argue about on a theoretical level here. At least it would require a more structured approach than what we can do in a GH thread, imho.

The bigger point I am worried about is that **if** this insertion-sorted `has_what` is what makes or breaks the rebalancing and we do not even see any way to work around this or evolve this to a more sophisticated pattern, that will severely limit our possibilities for the next steps. Once we move on to dynamic workloads, we most definitely need to drop the "last inserted" policy since the last inserted task will most likely always be a fetched dependency of another task and it will be required immediately. Moving this off to another worker would be a horrible decision. I do not even want to start the discussion about what the best or even a good policy will be for this step, I just want to point out that this will be likely much more complicated and will require more flexibility.

My entire point is that we can already anticipate this insertion sort to not be sufficient. My proposal to include an explicit sort here was one possibility to make the rebalance modular and evolvable, not necessarily scalable. IMHO, your proposal suggests a highly scalable algorithm which is less modular and difficult to evolve. I believe we are looking for something in between.